### PR TITLE
Add temporary directory cleanup workarounds

### LINKED ISSUE #31732: Take action to stop the temporary directory being deleted due to lack of use
Since #27609 Elasticsearch defaults to a per-run temporary directory.  On Linux this is a directory created by the startup script under `/tmp`.

Linux `systemd` has functionality that can remove files and directories from `/tmp` that have not been used for a certain length of time.  This functionality is described in `man tmpfiles.d`.

RHEL and CentOS 7 ship with a configuration for this functionality in `/usr/lib/tmpfiles.d/tmp.conf` that deletes files and directories that have not been touched for 10 days:

```
v /tmp 1777 root root 10d
```

(Note: If you read the man page you might think this _doesn't_ delete old files, as the man page says:

>       The age field only applies to lines starting with d, D, and x. If omitted or set to "-", no automatic clean-up is done.

However, the man page is wrong.  Cleanup by age also applies to other configuration entries, including v.  There are 7 letters it applies to in the code: https://github.com/systemd/systemd/blob/2479c4fe3fc3d0b631b93debbc2a83aa40a5f379/src/tmpfiles/tmpfiles.c#L1904

v is `CREATE_SUBVOLUME` in that `switch`.)

Currently the only part of Elasticsearch that uses `java.io.tmpdir` more than a few seconds after startup is ML.  As a result, if someone does not start an ML job on a particular node that is running on RHEL or CentOS 7 then 10 days after ES startup the temporary directory is removed by `tmpfiles.d` functionality.  If an ML job is run on the node after this then it fails because the temporary directory does not exist.

Due to security manager the ES JVM cannot recreate the temporary directory.  Therefore the best solution would seem to be to periodically create and remove a file in the temporary directory.  If we created and removed a file every 22 hours then this would keep the directory modification time within the last day, even for days when daylight saving time starting reduces the day length to 23 hours.  So this would keep the directory alive even for a user who configured `tmpfiles.d` to clean after 1 day.

Since ML is currently the only affected component this periodic touching of the temporary directory could be done in the ML code.  However, this problem could also affect 3rd party plugins that use `java.io.tmpdir`, so it would be nicer if the functionality to keep the temporary directory alive was in core Elasticsearch.  The ML team can implement it if we can get some advice on the best place in the code to put it.

### PR DESCRIPTION
**Pull request by droberts195**
On some Linux distributions tmpfiles.d cleans files and
directories under /tmp if they haven't been accessed for
10 days.

This can cause problems for ML as ML is currently the only
component that uses the temp directory more than a few
seconds after startup. If you didn't open an ML job for
10 days and then tried to open one then the temp directory
would have been deleted.

This commit prevents the problem occurring in the case of
Elasticsearch being managed by systemd, as systemd private
temp directories are not subject to periodic cleanup (by
default).

Additionally there are now some docs to warn people about
the risk and suggest a manual mitigation for .tar.gz users.

Fixes #31732

### CODE REVIEW

File: docs/reference/setup/important-settings/es-tmpdir.asciidoc [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+subsequently used.
+
+If you install Elasticsearch using the `.deb` or `.rpm` packages and run it
+under `systemd` then the private temporary directory that Elasticsearch uses
+is excluded from periodic cleanup.
+
+However, if you intend to run the `.tar.gz` distribution on Linux for an
+extended period then you should consider creating a dedicated temporary
+directory for Elasticsearch that is not under a path that will have old files
+and directories cleaned from it.  This directory should have permissions set
```
Comment (jasontedor) [inline docs/reference/setup/important-settings/es-tmpdir.asciidoc]: An incredibly minor nit, but for consistency: reduce these two spaces to one.

**Review by jasontedor — APPROVED** [review]
LGTM. I left a nit, no need for another round.

**Comment by jasontedor** [comment]
@elasticmachine test this please

**Comment by droberts195** [comment]
Looks like it's not as simple as I thought.  Elasticsearch itself happily started up under systemd with the extra setting when I tested this on a cloud server, but the packaging tests are consistently failing as some other part of the Elasticsearch script ecosystem isn't being redirected to the private directory.

**Comment by droberts195** [comment]
@elasticmachine run gradle build tests

**Comment by droberts195** [comment]
@elasticmachine run all packaging tests

**Pull request merged**